### ISSUE TITLE
vkCreateInstance "unbalanced the stack"

### ISSUE THREAD

**Issue by stefnotch** [jwollen/SharpVulkan#1]
Amazing project! I tried to get this to work in C# ( https://gist.github.com/graphitemaster/e162a24e57379af840d4 ) but this line:` instance = Vulkan.CreateInstance(ref instanceInfo);` doesn't work as intended. It says that `vkCreateInstance(__createInfo__, allocator, &instance).CheckError();` unbalanced the stack. (Should I post a screenshot?)

Here is my code:
```
using System;
using System.Runtime.InteropServices;
namespace SharpVulkan
{
    class Program
    {
        unsafe public static void Main(string[] args)
        {
            //also, you need to make sure to free the unmanaged memory:
            //Marshal.FreeHGlobal(strPtr);
            //Constants
            IntPtr NULL = IntPtr.Zero;

            ApplicationInfo applicationInfo = new ApplicationInfo();
            InstanceCreateInfo instanceInfo = new InstanceCreateInfo();
            Instance instance;
            
            // Filling out application description:
            // sType is mandatory
            applicationInfo.StructureType = StructureType.ApplicationInfo;
            // pNext is mandatory
            applicationInfo.Next = NULL;
            // The name of our application
            applicationInfo.ApplicationName = Marshal.StringToHGlobalUni("Tutorial 1");
            // The name of the engine (e.g: Game engine name)
            applicationInfo.EngineName = NULL;
            // The version of the engine
            applicationInfo.EngineVersion = 1;
            // The version of Vulkan we're using for this application
            applicationInfo.ApiVersion = Vulkan.ApiVersion;


            // Filling out instance description:
            // sType is mandatory
            instanceInfo.StructureType = StructureType.InstanceCreateInfo; // VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO;
            // pNext is mandatory
            instanceInfo.Next = NULL;
            // flags is mandatory
            instanceInfo.Flags = 0;
            // The application info structure is then passed through the instance

            // object to IntPtr (before calling WinApi):
            GCHandle handle1 = GCHandle.Alloc(applicationInfo);
            IntPtr parameter = (IntPtr)handle1;
            // call WinAPi and pass the parameter here

            //IntPtr x = NULL;
            //Marshal.StructureToPtr(applicationInfo, x, true); // &applicationInfo;
            instanceInfo.ApplicationInfo = parameter;
            // Don't enable and layer
            instanceInfo.EnabledLayerCount = 0;
            instanceInfo.EnabledLayerNames = NULL;
            // Don't enable any extensions
            instanceInfo.EnabledExtensionCount = 0;
            instanceInfo.EnabledExtensionNames = NULL;

            // Now create the desired instance
            instance = Vulkan.CreateInstance(ref instanceInfo);
            // To Come Later
            // ...
            // ...


            // Never forget to free resources
            instance.Destroy();

        }
    }
}```


Also, an example would be awesome!

**Comment by jwollen** [comment]
Hi, thanks for playing with it!

You will want to use a pinned handle:
```
GCHandle handle1 = GCHandle.Alloc(applicationInfo, GCHandleType.Pinned);
IntPtr parameter = handle1.AddrOfPinnedObject();
```
or better
```
instanceInfo.ApplicationInfo = new IntPtr(&applicationInfo);
```

That should work! However, I had no stack imbalance, but a Vulkan error, so let me know if you have more trouble.
Samples coming soon!

**Comment by stefnotch** [comment]
Thanks for the quick response! :D

Well, I tried out both variants and the same error popped up again:
Managed Debugging Assistant 'PInvokeStackImbalance' has detected a problem in 'C:\Stefan\HTL\SchoolServer\C#\Vulkan\Vulkan\bin\Debug\Vulkan.vshost.exe'.

Additional information: A call to PInvoke function 'SharpVulkan!SharpVulkan.Vulkan::vkCreateInstance' has unbalanced the stack. This is likely because the managed PInvoke signature does not match the unmanaged target signature. Check that the calling convention and parameters of the PInvoke signature match the target unmanaged signature.

(In case you are interested in my particular set up: https://drive.google.com/open?id=0B0p1M-iw1bBrTzJPY0V3YUYxYjg )
